### PR TITLE
Add metricdatatest package

### PR DESCRIPTION
**Pull request by MrAlias**
The `metricdata` package contains types that will need to be compared when testing output along the telemetry pipeline (e.g. aggregators, instrumentation libraries). This adds comparison and assertion functions for each data-type.

This is needed for #3000 and #3008

#### Use `github.com/stretchr/testify/assert`

There are types that have slices that need to be checked that the content matches, but not necessarily the order. The `assert.Equal` function uses `reflect.DeepEqual` under the hood which would not compare in this fashion.

Additionally, the `testify` error messages for complex type comparison can be hard to parse. This attempts to provided succinct and relevant reasons for failures given we have context of the the type.

### CODE REVIEW

**Review by MadVikingGod — COMMENTED** [review]
I have two issues with this:

1.  I don't know what the outputs of these Asserts are, so I have no way to compare them to the output of Assert.Equal().  Even if we don't store them in documentation (and immediately let it get out of date), I would like to see what different error messages would be.
2. the length of results seems to vary necessarily, but we are checking the length in some of our tests.  For example, AssertValues sometimes returns len()==1, and sometimes len()==2 but there is only one problem.  I woulder either only check that len>0, or try to make this more predictable (and documented).

One thing I saw that might simplify this is that it seems if len(reasons)>0 then equal is false.  I didn't check all cases, but could we drop the bool for a len check?

**Comment by MrAlias** [comment]
> One thing I saw that might simplify this is that it seems if len(reasons)>0 then equal is false. I didn't check all cases, but could we drop the bool for a len check?

That should be the case indeed, and that sounds like a good suggestion. I'll revise.

> 2. the length of results seems to vary necessarily, but we are checking the length in some of our tests.  For example, AssertValues sometimes returns len()==1, and sometimes len()==2 but there is only one problem.  I woulder either only check that len>0, or try to make this more predictable (and documented).

Correct. I didn't want to test directly the messages as they will likely change and it will add a lot of code to the testing. I can switch to check the length or the reasons is greater than zero, but I was worried it will miss expected results (i.e. something returns errors for two fields but not the third).

I do think the testing of these things can be iterated in the future if we find it necessary. I'll switch to just checking the length of the response is greater than zero.

> After posting that I came up with a way of running the failing tests without breaking CI. Here's the test file with a build tag, and the output: https://gist.github.com/MadVikingGod/e6bd791c2747d9311b936ab31a6b1cdb

Thanks for including this. I didn't know the best way to add this failure info to the description, this is helpful.

> After looking at the output, a few things stick out. Some of these are easy to tell what's wrong, but some less so: In the ResourceMetrics test: resources are easy to tell, but the scopemetrics less so:

For the ScopeMetrics, what would be a clearer way to you to see they are not the same for the ResourceMetric? I am trying to concisely state that the elements of the ScopeMetric slice differ, and showing what one contains but not the other, but I agree there is a lot of information there.

File: sdk/metric/metricdata/metricdatatest/comparisons.go [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+	}
+	if a.Description != b.Description {
+		equal, reasons = false, append(reasons, notEqualStr("Description", a.Description, b.Description))
+	}
+	if a.Unit != b.Unit {
+		equal, reasons = false, append(reasons, notEqualStr("Unit", a.Unit, b.Unit))
+	}
+
+	var r []string
+	equal, r = equalAggregations(a.Data, b.Data)
```
Comment (MadVikingGod) [inline sdk/metric/metricdata/metricdatatest/comparisons.go:96]: We are overwriting `equal` in all of these cases.  if the aggregations are the same but the name is different then this will return true.

Comment (MrAlias) [reply sdk/metric/metricdata/metricdatatest/comparisons.go]: Resolved now that equalAggregations no longer returns a bool.

**Comment by MadVikingGod** [comment]
> > After looking at the output, a few things stick out. Some of these are easy to tell what's wrong, but some less so: In the ResourceMetrics test: resources are easy to tell, but the scopemetrics less so:
> 
> For the ScopeMetrics, what would be a clearer way to you to see they are not the same for the ResourceMetric? I am trying to concisely state that the elements of the ScopeMetric slice differ, and showing what one contains but not the other, but I agree there is a lot of information there.

I think there are two solutions that would help:

1. Add formatting to the objects, like either a `String() string` function or some other formating, so we have just a bit less extra information.
2. Use a formatting function that is aware of the different types we have and can provide a more concise format.

My goal would be if presented with something like:
```
            ResourceMetrics ScopeMetrics not equal:
            missing expected values:
            metricdata.ScopeMetrics{Scope:instrumentation.Scope{Name:"A", Version:"", SchemaURL:""}, Metrics:[]metricdata.Metrics{metricdata.Metrics{Name:"A", Description:"A desc", Unit:"1", Data:metricdata.Sum{DataPoints:[]metricdata.DataPoint{metricdata.DataPoint{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[1]attribute.KeyValue{attribute.KeyValue{Key:"A", Value:attribute.Value{vtype:1, numeric:0x1, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2022, time.July, 20, 13, 18, 7, 260802950, time.Local), Time:time.Date(2022, time.July, 20, 13, 18, 8, 260802950, time.Local), Value:-1}}, Temporality:0x1, IsMonotonic:true}}}}
            unexpected additional values:
            metricdata.ScopeMetrics{Scope:instrumentation.Scope{Name:"B", Version:"", SchemaURL:""}, Metrics:[]metricdata.Metrics{metricdata.Metrics{Name:"B", Description:"B desc", Unit:"By", Data:metricdata.Gauge{DataPoints:[]metricdata.DataPoint{metricdata.DataPoint{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[1]attribute.KeyValue{attribute.KeyValue{Key:"B", Value:attribute.Value{vtype:1, numeric:0x1, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2022, time.July, 20, 13, 18, 7, 261802950, time.Local), Time:time.Date(2022, time.July, 20, 13, 18, 8, 261802950, time.Local), Value:2}}}}}}
  ``` 
and it only differed by one character I would be easy to tell apart.

**Review by MadVikingGod — APPROVED** [review]
I think the messages can use a bit of work to make them more readable, but that can be done as a revision

**Comment by MrAlias** [comment]
> I think the messages can use a bit of work to make them more readable, but that can be done as a revision

Agreed. I started in on a refactor to improve them but it will add a lot to this PR. I'll merge this now as it has the approvals and open an issue to address the message refactor.

**Comment by MrAlias** [comment]
https://github.com/open-telemetry/opentelemetry-go/issues/3035

**Comment by MrAlias** [comment]
> I'll merge this now as it has the approvals and open an issue to address the message refactor.

Err, never mind. Still need another review.

**Review by dashpole — APPROVED** [review]
